hbase-client/src/main/java/org/apache/hadoop/hbase/client/AsyncNonMetaTableRegionLocator.java

```diff
@@ -0,0 +1,148 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.hadoop.hbase.client;
+
+import java.io.IOException;
+import java.util.List;
+import java.util.concurrent.CompletableFuture;
+import org.apache.hadoop.hbase.AsyncMetaTableAccessor;
+import org.apache.hadoop.hbase.HRegionLocation;
+import org.apache.hadoop.hbase.MetaTableAccessor;
+import org.apache.hadoop.hbase.RegionLocations;
+import org.apache.hadoop.hbase.TableName;
+import org.apache.hadoop.hbase.TableNotFoundException;
+import org.apache.hadoop.hbase.util.Bytes;
+import org.apache.yetus.audience.InterfaceAudience;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * The class for locating region for table other than meta.
+ */
+@InterfaceAudience.Private
+class AsyncNonMetaTableRegionLocator extends AbstractAsyncTableRegionLocator {
```

Comment:
Two locators. Can we not have one?

Reply:
An AsyncNonMetaTableRegionLocator is for user-space Regions? We also have AsyncRegionLocator... and Abstract* Do we have to have this many classes?

Reply:
We do have these locators before this patch. Th AsyncRegionLocator is for the general interface to communicate with upper layer, and in the past, we have AsyncMetaRegionLocator and AsyncNonMetaRegionLocator, for dealing with meta table and non meta table.

In this patch, the old stuff in AsyncMetaRegionLocator has all been purged, and most of the logic in AsyncNonMetaRegionLocator have been moved to AsyncRegionLocator and AbstractAsyncTableRegionLocator. The latter one is created for sharing code between meta table and non meta table.

We can put all these logics in a single class, as all class inherits can also be done by a if in the code. But I think doing this is an anti-pattern in java, as it will introduce very big classes, which is already a big problem in HBase(think of the old AsyncProcess class, it is a nightmare to me...).

Reply:
Ok. I'm reacting to our Interface => Abstract => Impl tiering.... We do it too often.